docs/blog/2021-12-01-csp.md

```diff
@@ -0,0 +1,153 @@
+```html
+<script src="https://accounts.google.com/o/oauth2/revoke?callback=alert('This is a malicious alert')"></script>
+```
+Will return this:
+```javascript
+// API callback
+alert('This is a malicious alert')({
+  // ...googles intended payload
+}
+);
+```
+
+Go ahead and try [the url from that script](<https://accounts.google.com/o/oauth2/revoke?callback=alert('This is a malicious alert')>) in your browser.
+You'll quickly see the problem.  Now an attacker that can embed HTML in your page has
+the ability to execute whatever javascript they want. Any un-sanitized field that leaks HTML into your page is now an immediate vulnerability.
+
+There a great number of these vulnerable legacy endpoints, not just on googles domains but across the web.  
+Here is a list of [just some of them](https://github.com/zigoo0/JSONBee/blob/master/jsonp.txt).
+
+Because this isn't a specific protocol but just a serverside (anti)pattern, there's no way for the browser
+to detect this behavior and disable it.  The only solution is to not allow scripts to load from any vulnerable domain, and 
+to host them yourself instead.
+
+ **The bigger your website, the harder this is to deal with.**
+
+If you have a small site and immediate control over the codebase, this could be an easy problem to patch.
+
+For large sites at big companies that load analytics payloads, third party auth, and other scripts from these vulnerable domains, it can be 
+difficult to mitigate.  Heck, a lot of big companies have legacy JSONP endpoints on the _same domain as the site._  There's not much you can do about that.
+
+### File Upload Bypass
+Let's say your CSP is locked to only allow scripts from your domain.  
+
+If your site allows file uploads and downloads, including files that end in `.js`, a user can simply upload a file `malicious-code.js`,
+and load it as a script tag, directly from your domain. 
+
+```html
+<script src=www.yourdomain.com/myfiles/malicious-code.js></script>
+```
+
+You have to be very careful about what your users can upload.  
+
+### Other vulnerabilities
+These are just a few of the potential ways to bypass the CSP's protections.
+The list [goes on](https://book.hacktricks.xyz/pentesting-web/content-security-policy-csp-bypass).
+
+### Mitigations
+For large sites, there may be just too many chinks in the armor to treat the CSP as a functional line of defence, and many big sites have disabled it entirely. 
+[This paper](https://research.google/pubs/pub45542/) found that of those that use it **over 94% of sites have ineffective CSPs!**
+
+Long story short, there's a good chance your CSP isn't doing much to protect your site from XSS.
+
+Nonce-based CSP strategies can help mitigate the above attacks, where the server adds a random nonce attribute to every script tag and defines in the CSP.
+Embedding a script tag smart enough to clone the nonce isn't possible without javascript.  
+
+Unfortunately this strategy is the most work to implement and is rarely used. 
```

Comment:
I'd probably mention that the increasingly popular JAM stack pattern of serving documents off of a CDN makes this difficult because the Nonce must be random for it to be secure.

...

There really aren't very many resource available online to talk about this! Only this Tweet[0]. It probably warrants a separate blog post where we can do a deep dive into "How to make your Jamstack app secure with CSP". In order to _actually_ do this, you have to use something like `Lambda@Edge` or CloudFront Workers to rewrite the response (see this[1] post).

Basically, choose either a Nonce or a low-latency CDN. (Or the middle ground of writing code to rewrite the cached response)

0: https://twitter.com/mikesherov/status/1378373386171191299

1: https://medium.com/@csarkosh/my-experience-getting-an-a-from-mozillas-observatory-tool-on-aws-f0abf12811a1